data_iterator.go

```diff
@@ -107,81 +130,135 @@ func (d *DataIterator) Run(tables []*TableSchema) {
 						}
 
-						batch = &RowBatch{
-							values:             rows,
-							paginationKeyIndex: batch.PaginationKeyIndex(),
-							table:              table,
-							fingerprints:       fingerprints,
-						}
+						batch.values = rows
+						batch.fingerprints = fingerprints
 					}
 
 					for _, listener := range d.batchListeners {
 						err := listener(batch)
 						if err != nil {
-							logger.WithError(err).Error("failed to process row batch with listeners")
+							batchLogger.WithError(err).Error("failed to process row batch with listeners")
 							return err
 						}
+
+						if d.gracefulShutdown {
+							return GracefulShutdownError{}
+						}
 					}
 
 					return nil
 				})
 
 				if err != nil {
 					switch e := err.(type) {
+					case GracefulShutdownError:
+						d.logger.Warn("received gracefully shutdown signal")
+						return
 					case BatchWriterVerificationFailed:
-						logger.WithField("incorrect_tables", e.table).Error(e.Error())
+						d.logger.WithField("incorrect_tables", e.table).Error(e.Error())
 						d.ErrorHandler.Fatal("inline_verifier", err)
 					default:
-						logger.WithError(err).Error("failed to iterate table")
+						d.logger.WithError(err).Error("failed to iterate table")
 						d.ErrorHandler.Fatal("data_iterator", err)
 					}
 
 				}
-
-				logger.Debug("table iteration completed")
-
-				// Right now the BatchWriter.WriteRowBatch happens synchronously in
-				// this method. If it ever becomes async, this MarkTableAsCompleted
-				// call MUST be done in WriteRowBatch somehow.
-				d.StateTracker.MarkTableAsCompleted(table.String())
 			}
 		}()
 	}
 
-	i := 0
-	loggingIncrement := len(tablesWithData) / 50
-	if loggingIncrement == 0 {
-		loggingIncrement = 1
-	}
+	for _, key := range d.paginationKeys {
+		tableName := key.table.String()
+		stateBatches, loadBatchesFromState := d.StateTracker.batchProgress[tableName]
 
-	for table, _ := range tablesWithData {
-		tablesQueue <- table
-		i++
-		if i%loggingIncrement == 0 {
-			d.logger.WithField("table", table.String()).Infof("queued table for processing (%d/%d)", i, len(tablesWithData))
+		if loadBatchesFromState {
+			for _, batch := range stateBatches {
+				if batch.Completed {
+					continue
+				}
+
+				batchQueue <- &PaginationKey{
+					table:              key.table,
+					startPaginationKey: batch.LatestPaginationKey,
+					endPaginationKey:   batch.EndPaginationKey,
+				}
+			}
+		} else {
+			// Set start to minus one since cursor is searching for greater values
+			tableStartPaginationKey := key.startPaginationKey - 1
+			tableEndPaginationKey := key.endPaginationKey
+
+			// Number of batches are set to number of processes, unless each batch becomes smaller than the cursor size
+			keyInterval := tableEndPaginationKey - tableStartPaginationKey
+			concurrencyBatchSize := math.Ceil(float64(keyInterval) / float64(d.Concurrency))
+			batchSize := uint64(math.Max(concurrencyBatchSize, float64(d.CursorConfig.BatchSize)))
```

Comment:
Would like to see some tests for this

Reply:
I actually went and reverted all test changes from this branch for now. Will go back and figure out how to test this properly since it need some changes to how we assert and halt on iteration loops etc.